src/neo/Network/P2P/Payloads/Signer.cs

```diff
@@ -35,17 +36,22 @@ public class Signer : ISerializable
         /// </summary>
         public ECPoint[] AllowedGroups;
 
+        public IDictionary<UInt160, UInt160[]> AllowedCallingContracts;
+        public IDictionary<UInt160, ECPoint[]> AllowedCallingGroup;
```

Comment:
Why not use a single array?

Reply:
To use it more efficiently later:
https://github.com/Ashuaidehao/neo/blob/5015c4f7268a5dc8e7b9f2f3c3f119914182be17/src/neo/SmartContract/ApplicationEngine.Runtime.cs#L230

Reply:
I think it should be:

```c#
public UInt160[] AllowedCallingContracts;
public ECPoint[] AllowedCallingGroup;
```

Reply:
> I think it should be:
> 
> ```cs
> public UInt160[] AllowedCallingContracts;
> public ECPoint[] AllowedCallingGroup;
> ```

```
CustomCallingGroups set contract with trusted caller's group list. Only Trust Group Caller or entry call can pass "CheckWitness".

{
	TokenA:[Group1,Group2,...],
	TokenB:[Group1,Group3,...],
	...
}
```

key-value structure needed.